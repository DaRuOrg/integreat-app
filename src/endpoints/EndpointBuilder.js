import { isEqual } from 'lodash/lang'

import Endpoint from './Endpoint'
import StateMapperBuilder from './StateMapperBuilder'

/**
 * Helper class to build a {@link Endpoint}
 */
class EndpointBuilder {
  _name
  _url
  _mapper
  _stateMapperBuilder
  _refetchLogic

  /**
   * Creates a new endpoint builder
   * @param {string} name The name of the endpoint to build
   */
  constructor (name) {
    this._name = name
    this._stateMapperBuilder = new StateMapperBuilder(this)

    this._refetchLogic = (urlParams, nextUrlParams) => !isEqual(urlParams, nextUrlParams)
  }

  /**
   * Adds a url to the builder
   * @param url The url
   * @return {EndpointBuilder} The builder itself
   */
  withUrl (url) {
    this._url = url
    return this
  }

  /**
   * Adds a url to the builder
   * @param mapper The mapper
   * @return {EndpointBuilder} The builder itself
   */
  withMapper (mapper) {
    this._mapper = mapper
    return this
  }

  /**
   * Adds a state mapper builder to this builder
   * @return {StateMapperBuilder} The state mapper of this builder
   */
  withStateMapper () {
    return this._stateMapperBuilder
  }

  /**
   * Adds refetch logic to this builder
   * @param refetchLogic The refetch logic
   * @return {EndpointBuilder}  The builder itself
   */
  withRefetchLogic (refetchLogic) {
    this._refetchLogic = refetchLogic
    return this
  }

  /**
   * Checks the data and builds the endpoint
   * @return {Endpoint} The final endpoint
   */
  build () {
    if (!this._name) {
      throw Error('You have to set a name to build an endpoint!')
    }

    if (!this._url) {
      throw Error('You have to set a url to build an endpoint!')
    }

    if (!this._mapper) {
      throw Error('You have to set a mapper to build an endpoint!')
    }

    if (!this._url) {
      throw Error('You have to set a url to build an endpoint!')
    }

    if (!this._stateMapperBuilder) {
      throw Error('You have to set a state mapper to build an endpoint!')
    }

    if (!this._refetchLogic) {
      throw Error('You have to set a refetch logic to build an endpoint!')
    }

    return new Endpoint(this._name, this._url, this._mapper, this._stateMapperBuilder.build(), this._refetchLogic)
  }
}

export default EndpointBuilder
