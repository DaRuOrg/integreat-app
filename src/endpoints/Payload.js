class Payload {
  constructor (isFetching, data = null, error = null, requestUrl = null) {
    this._isFetching = isFetching
    this._data = data
    this._error = error
    this._requestUrl = requestUrl
  }

  get isFetching () {
    return this._isFetching
  }

  get data () {
    return this._data
  }

  ready () {
    return !!this._data
  }

  get error () {
    return this._error
  }

  get requestUrl () {
    return this._requestUrl
  }
}

export default Payload
