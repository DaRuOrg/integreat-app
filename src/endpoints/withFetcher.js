import React from 'react'
import cx from 'classnames'
import { connect } from 'react-redux'
import Spinner from 'react-spinkit'

import Error from 'components/Error'
import style from './Fetcher.css'

function createStateToPropsMapper (endpoint) {
  return (state) => ({
    [endpoint.payloadName]: state[endpoint.stateName],
    options: endpoint.mapStateToOptions(state)
  })
}

function withFetcher (endpoint, hideError = false, hideSpinner = false) {
  return (WrappedComponent) => {
    let Fetcher = class extends React.Component {
      static displayName = endpoint.name + 'Fetcher'

      constructor () {
        super()
        this.state = { isDataAvailable: false }
      }

      fetch (options) {
        if (!options) {
          throw new Error('options are not valid! This could mean your mapStateToOptions() returns ' +
            'a undefined value!')
        }
        const isDataAvailable = this.props.dispatch(endpoint.requestAction(options, options))
        this.setIsDataAvailable(isDataAvailable)
      }

      setIsDataAvailable (isDataAvailable) {
        this.setState({ isDataAvailable })
      }

      componentWillMount () {
        // We need to have this discussion in mind, when building the fetcher architecture,
        // because a store dispatch in the componentWillMount has no immediate effect on the props from connect() for
        // the first call of render() (and therefore the <WrappedComponent> would have been mounted for one moment, if
        // we just checked to the payload.ready() prop)
        // https://github.com/reactjs/react-redux/issues/210#issuecomment-166055644
        this.fetch(this.props.options)
      }

      componentWillUpdate (nextProps) {
        // Dispatch new RequestAction to ask the endpoint whether the fetcher can display the data, if
        // (a) the endpoint properties change or
        // (b) the fetcher receives new payload information from the store (e.g. because a payload has been fetched)
        if (endpoint.shouldRefetch(this.props.options, nextProps.options) || this.props[endpoint.payloadName] !== nextProps[endpoint.payloadName]) {
          this.fetch(nextProps.options)
        }
      }

      errorVisible () {
        return !hideError && this.props[endpoint.payloadName].error
      }

      render () {
        const payload = this.props[endpoint.payloadName]

        if (!this.state.isDataAvailable) {
          if (!hideSpinner) {
            return <Spinner className={cx(style.loading, this.props.className)} name='line-scale-party'/>
          } else {
            return <div/>
          }
        }

        if (this.errorVisible()) {
          return <Error className={cx(style.loading, this.props.className)} error={payload.error}/>
        }

        return <WrappedComponent {...Object.assign({}, this.props, {[endpoint.stateName]: payload.data})}/>
      }
    }

    return connect(createStateToPropsMapper(endpoint))(Fetcher)
  }
}

export default withFetcher
