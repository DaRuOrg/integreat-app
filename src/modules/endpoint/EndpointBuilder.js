import { isEqual } from 'lodash/lang'

import Endpoint from './Endpoint'

/**
 * Helper class to build a {@link Endpoint}
 */
class EndpointBuilder {
  _name
  _routerToUrlMapper
  _mapper
  _responseOverride
  _refetchLogic

  /**
   * Creates a new endpoint builder
   * @param {string} name The name of the endpoint to build
   */
  constructor (name) {
    this._name = name

    this._refetchLogic = (urlParams, nextUrlParams) => !isEqual(urlParams, nextUrlParams)
  }

  /**
   * Adds a router to url mapper to the builder
   * @param routerToUrlMapper The url
   * @return {EndpointBuilder} The builder itself
   */
  withRouterToUrlMapper (routerToUrlMapper) {
    this._routerToUrlMapper = routerToUrlMapper
    return this
  }

  /**
   * Adds a url to the builder
   * @param mapper The mapper
   * @return {EndpointBuilder} The builder itself
   */
  withMapper (mapper) {
    this._mapper = mapper
    return this
  }

  /**
   * Adds refetch logic to this builder
   * @param refetchLogic The refetch logic
   * @return {EndpointBuilder} The builder itself
   */
  withRefetchLogic (refetchLogic) {
    this._refetchLogic = refetchLogic
    return this
  }

  /**
   * Overrides value from the API response. Useful for testing.
   * @param responseOverride {*} The response
   * @return {EndpointBuilder} The builder itself
   */
  withResponseOverride (responseOverride) {
    this._responseOverride = responseOverride
    return this
  }

  /**
   * Checks the data and builds the endpoint
   * @return {Endpoint} The final endpoint
   */
  build () {
    if (!this._name) {
      throw Error('You have to set a name to build an endpoint!')
    }

    if (!this._routerToUrlMapper) {
      throw Error('You have to set a url to build an endpoint!')
    }

    if (!this._mapper) {
      throw Error('You have to set a mapper to build an endpoint!')
    }

    if (!this._refetchLogic) {
      throw Error('You have to set a refetch logic to build an endpoint!')
    }

    return new Endpoint(this._name, this._routerToUrlMapper, this._mapper, this._refetchLogic, this._responseOverride)
  }
}

export default EndpointBuilder
