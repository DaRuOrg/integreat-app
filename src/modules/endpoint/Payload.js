/**
 * The payload gets stored in the redux store and holds the information about a etch
 */
class Payload {
  constructor (isFetching, data = null, error = null, requestUrl = null, fetchDate = null) {
    this._isFetching = isFetching
    this._fetchDate = fetchDate
    this._data = data
    this._error = error
    this._requestUrl = requestUrl
  }

  /**
   * @return {number} The date the fetch was initiated as serializable number
   */
  get fetchDate () {
    return this._fetchDate
  }

  /**
   * @return {boolean} If a fetch is going on
   */
  get isFetching () {
    return this._isFetching
  }

  /**
   * @return {*} The data which has been fetched or null
   */
  get data () {
    return this._data
  }

  /**
   * @return {boolean} If the {@link data} is ready to be used
   */
  ready () {
    return !!this._data
  }

  /**
   * @return {string} The error message if the fetch failed or null
   */
  get error () {
    return this._error
  }

  /**
   * @return {string} The url which was used to initiate the fetch
   */
  get requestUrl () {
    return this._requestUrl
  }
}

export default Payload
