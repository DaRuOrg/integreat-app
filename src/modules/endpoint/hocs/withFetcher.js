import React from 'react'
import PropTypes from 'prop-types'
import cx from 'classnames'
import { connect } from 'react-redux'
import Spinner from 'react-spinkit'

import Error from 'modules/common/containers/Error'
import style from './withFetcher.css'
import { getContext } from 'recompose'

const contextTypes = {
  getEndpoint: PropTypes.func.isRequired
}

/**
 * This function builds a HOC from a component
 * @callback buildHOC
 * @param {React.Component} WrappedComponent
 * @return {React.Component} The HOC
 */

/**
 * This creates a factory for a Higher-Order-Component. The HOC attaches a fetcher to the supplied component.
 * @param endpointName {string} The name of the endpoint to fetch from
 * @param hideError {boolean} If you want to hide errors in the render() method
 * @param hideSpinner {boolean} If you want to hide a loading spinner in the render() method
 * @return {buildHOC} Returns a HOC which renders the supplied component as soon as the fetcher succeeded
 */
export function withFetcher (endpointName, hideError = false, hideSpinner = false) {
  return (WrappedComponent) => {
    class Fetcher extends React.Component {
      static displayName = endpointName + 'Fetcher'
      static propTypes = {
        urlParams: PropTypes.objectOf(PropTypes.string),
        className: PropTypes.string,
        requestAction: PropTypes.func.isRequired,
        getEndpoint: PropTypes.func
      }

      static contextTypes = contextTypes

      constructor (props, context) {
        super()
        this.state = {isDataAvailable: false}

        if (props.getEndpoint) {
          this.endpoint = props.getEndpoint(endpointName)
        } else {
          this.endpoint = context.getEndpoint(endpointName)
        }
      }

      componentWillMount () {
        // A store dispatch in the componentWillMount (and therefore a state update) does not cause the props to update
        // before render() in the Fetcher component is called the first time.
        // This causes the <WrappedComponent> to be displayed with outdated props.
        // https://github.com/reactjs/react-redux/issues/210#issuecomment-166055644
        this.fetch(this.props.urlParams)
      }

      componentWillReceiveProps (nextProps) {
        // Dispatch new requestAction to ask the endpoint whether data is available, if:
        // (a) the Fetcher urlParams prop changed or
        // (b) the Fetcher endpoint.payloadName prop changed because of new data in the store (e.g. because a payload has been fetched)
        const endpoint = this.endpoint
        if (endpoint.shouldRefetch(this.props.urlParams, nextProps.urlParams) ||
          this.props[endpoint.payloadName] !== nextProps[endpoint.payloadName]) {
          this.fetch(nextProps.urlParams)
        }
      }

      /**
       * Triggers a new fetch if available
       * @param {object} urlParams The params to use
       */
      fetch (urlParams) {
        if (!urlParams) {
          throw new Error('urlParams are not valid! This could mean your mapStateToUrlParams() returns ' +
            'a undefined value!')
        }
        const storeResponse = this.props.requestAction(urlParams)
        this.setState({isDataAvailable: storeResponse.dataAvailable})
      }

      errorVisible () {
        return !hideError && this.props[this.endpoint.payloadName].error
      }

      render () {
        const payload = this.props[this.endpoint.payloadName]

        if (!this.state.isDataAvailable) {
          if (!hideSpinner) {
            return <Spinner className={cx(style.loading, this.props.className)} name='line-scale-party' />
          } else {
            return <div />
          }
        }

        if (this.errorVisible()) {
          return <Error className={cx(style.loading, this.props.className)} error={payload.error} />
        }

        return <WrappedComponent {...Object.assign({}, this.props, {[this.endpoint.stateName]: payload.data})} /> //todo pass only props which are needed, no internal data
      }
    }

    return Fetcher
  }
}

const createMapStateToProps = (endpointName) => (state, ownProps) => {
  const endpoint = ownProps.getEndpoint(endpointName)
  return ({
    [endpoint.payloadName]: state[endpoint.stateName],
    urlParams: endpoint.mapStateToUrlParams(state)
  })
}

const createMapDispatchToProps = (endpointName) => (dispatch, ownProps) => {
  const endpoint = ownProps.getEndpoint(endpointName)
  return ({
    requestAction: (urlParams) => dispatch(endpoint.requestAction(urlParams))
  })
}

export default (endpointName, hideError, hideSpinner) => {
  const HOC = withFetcher(endpointName, hideError, hideSpinner)
  return (WrappedComponent) => {
    const AnotherWrappedComponent = HOC(WrappedComponent)
    return getContext(contextTypes)(
      connect(createMapStateToProps(endpointName), createMapDispatchToProps(endpointName))(AnotherWrappedComponent))
  }
}
