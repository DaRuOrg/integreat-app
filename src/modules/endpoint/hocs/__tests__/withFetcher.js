import React from 'react'
import withFetcher from '../withFetcher'
import { Provider } from 'react-redux'
import { mount } from 'enzyme'
import configureMockStore from 'redux-mock-store'
import EndpointBuilder from '../../EndpointBuilder'
import thunk from 'redux-thunk'
import Payload from '../../Payload'

describe('withFetcher', () => {
  const mockStore = configureMockStore([thunk])
  const endpoint = new EndpointBuilder('endpoint')
    .withUrl('https://someendpoint/api.json')
    .withMapper((json) => json)
    .withResponseOverride({test: 'random'})
    .build()

  test('should ', () => {
    const store = mockStore({endpoint: new Payload(false)})
    const TestComponent = () => <span>Text</span>
    const HOC = withFetcher(endpoint, true, true)
    const Hoced = HOC(TestComponent)

    const mounted = mount(<Provider store={store}>
      <Hoced/>
    </Provider>)


  })
})
