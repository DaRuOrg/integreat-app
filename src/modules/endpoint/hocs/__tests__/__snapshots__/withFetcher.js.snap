// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withFetcher should should show error if there is one and it's not hidded 1`] = `
<Connect(Component)
  className="loading"
  error="Yepp... Error time! Wuschhh!"
/>
`;
